website/blog/2024-04-26-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,287 @@
+[PEFT module](https://huggingface.co/blog/peft) enhances productivity and scalability by reducing code complexity and enabling the use of
+high-quality pre-optimized models for a range of diverse NLP tasks.
+
+Below is an example of configuring the PEFT-based fine tuning of a base model using PyTorch Lightning and HuggingFace's `peft` module.
+
+```python
+from typing import List
+from lightning import LightningModule
+from peft import get_peft_model, LoraConfig, TaskType
+from transformers import AutoModelForSequenceClassification
+
+
+class TransformerModule(LightningModule):
+    def __init__(
+        self,
+        pretrained_model: str,
+        num_classes: int = 2,
+        lora_alpha: int = 32,
+        lora_dropout: float = 0.1,
+        r: int = 8,
+        lr: float = 2e-4
+    ):
+        super().__init__()
+        self.model = self.create_model(pretrained_model, num_classes, lora_alpha, lora_dropout, r)
+        self.lr = lr
+        self.save_hyperparameters("pretrained_model")
+
+    def create_model(self, pretrained_model, num_classes, lora_alpha, lora_dropout, r):
+        """Create and return the PEFT model with the given configuration.
+
+        Args:
+            pretrained_model: The path or identifier for the pretrained model.
+            num_classes: The number of classes for the sequence classification.
+            lora_alpha: The alpha parameter for LoRA.
+            lora_dropout: The dropout rate for LoRA.
+            r: The rank of LoRA adaptations.
+
+        Returns:
+            Model: A model configured with PEFT.
+        """
+        model = AutoModelForSequenceClassification.from_pretrained(
+            pretrained_model_name_or_path=pretrained_model,
+            num_labels=num_classes
+        )
+        peft_config = LoraConfig(
+            task_type=TaskType.SEQ_CLS,
+            inference_mode=False,
+            r=r,
+            lora_alpha=lora_alpha,
+            lora_dropout=lora_dropout
+        )
+        return get_peft_model(model, peft_config)
+
+    def forward(self, input_ids: List[int], attention_mask: List[int], label: List[int]):
+        """Calculate the loss by passing inputs to the model and comparing against ground truth labels.
+
+        Args:
+            input_ids: List of token indices to be fed to the model.
+            attention_mask: List to indicate to the model which tokens should be attended to, and which should not.
+            label: List of ground truth labels associated with the input data.
+
+        Returns:
+            torch.Tensor: The computed loss from the model as a tensor.
+        """
+        return self.model(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            labels=label
+        )
+```
+
+Additional references for the full implementation can be [seen within the companion repository here](https://github.com/puneet-jain159/deeplearning_with_mlfow/blob/master/custom_module/fine_tune_clsify_head.py)
```

Comment:
nit: not related to the blog content, but the linked code has a bit incorrect Python syntax
https://github.com/puneet-jain159/deeplearning_with_mlfow/blob/master/custom_module/fine_tune_clsify_head.py#L60-L100

There are two `return` in the `predict` function where anything below L77 is simply discarded.

Reply:
Hi @B-Step62  
Corrected the code.  and tested the updated code with an end 2 end run. 

Link : https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/4093282202515276